// poetry/lib/db/then.js
const Database = require( './connection.js' ),
    //Events = require( '../methods/events' ),
    Log = require( '../methods/log' ),

    /**
     * After functions
     * @param {mixed} result - query result
     * @param {object} db - database connection
     * @param {Array} args - arguments
     * @returns {object} promise (only handle resolve)
     */
    After = {
        'insert': require( './after/insertSaveSet.js' ),
        'save': require( './after/insertSaveSet.js' ),
        'set': require( './after/insertSaveSet.js' ),
        'update': require( './after/update.js' )
    };

var service = 'unknown service';
const DEBUG = process.env.debug;

try {
    var pkg = require( process.cwd() + '/package.json' );
    service = pkg.name || service;
} catch ( e ) {}

module.exports = ( target, collection, action, args ) =>
    function queryThen( cb, err ) {
        const uuid = Date.now() + Math.random();
        if(DEBUG) console.log('['+uuid+'] QueryTHEN');

        let p = new Promise( ( resolve, reject ) => {

            Database( collection )
                .then( database => {
                    if ( !database ) {
                        console.error('NO DATABASE');
                        return resolve(undefined);
                    }

                    if ( !database[ action ] )
                        return reject(
                            `Method '${action}' not found for '${collection}'`
                        );

                    if(DEBUG) Log.silly('['+uuid+'] DATABASE CALL', collection, action, JSON.stringify(args));


                    let result = database[ action ]
                        .apply( database, args );

                        if(DEBUG) Log.silly('['+uuid+'] CALL SENT TO DB');

                    while ( target.query.length ) {
                        q = target.query.shift();
                        if ( result[ q.action ] )
                            result = result[ q.action ]
                            .apply( result, q.args );
                        else
                            return reject(
                                `Chained-method '${q.action}' not found`
                            );
                    }

                    if ( result.toArray )
                        result = result.toArray();

                        if(DEBUG) Log.silly('['+uuid+'] RECEIVED DB CURSOR');

                    if ( !result.then )
                        return customcb( result );

                    return result.then( customcb, (a) => {
                        if(DEBUG) Log.error('['+uuid+'] DB REJECTED', a);
                        reject(a);
                    });

                    function customcb( result ) {

                        if(DEBUG) Log.silly('['+uuid+'] DB RESOLVED', JSON.stringify(result));

                        if ( !result ) return resolve( result );

                        if ( After[ action ] ) {
                            if(DEBUG) Log.silly('['+uuid+'] AFTER DB EXEC');
                            After[ action ]( result, database, args )
                                .then(
                                    ( res ) => {
                                        if(DEBUG) Log.silly('['+uuid+'] DB DELEGATE 1');
                                        resolve( res );
                                        emitEvent(
                                            action,
                                            collection,
                                            res
                                        );
                                    }
                                );
                        } else {
                            if ( result.ops )
                                result = result.ops;

                            if ( result.value )
                                result = result.value;

                            if(DEBUG) Log.silly('['+uuid+'] DB DELEGATE 2');
                            resolve( result );
                            emitEvent( action,
                                collection,
                                result
                            );
                        }
                    }

                }, reject );
        } );

        if(DEBUG) Log.silly('['+uuid+'] DB CB');

        if ( cb && err )
            p.then( cb, err );
        else if ( cb )
            p.then( cb );

        return p;
    };

function emitEvent( ev, collectionName, res ) {
    ev = ( ev === 'findAndModify' ) ? 'update' : ev;
    if ( collectionName == "poetry_events" ) return;

    if ( [
            'insert',
            'update',
            'save'
        ].indexOf( ev ) > -1 ){
            Database( 'poetry_events' )
                .then(db => {
                    db.insert({
                        name: ev + ':' + collectionName,
                        message: [res],
                        service: service,
                        createdAt: new Date()
                    })
                    .then(() => { }, Log.error);
                }, Log.error);
        }
}
